pinot-plugins/pinot-input-format/pinot-protobuf/src/main/java/org/apache/pinot/plugin/inputformat/protobuf/ProtoBufMessageDecoder.java

```diff
@@ -0,0 +1,105 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.plugin.inputformat.protobuf;
+
+import com.github.os72.protobuf.dynamic.DynamicSchema;
+import com.google.common.base.Preconditions;
+import com.google.protobuf.Descriptors;
+import com.google.protobuf.DynamicMessage;
+import com.google.protobuf.Message;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URI;
+import java.util.Map;
+import java.util.Set;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.pinot.spi.data.readers.GenericRow;
+import org.apache.pinot.spi.stream.StreamMessageDecoder;
+import org.apache.pinot.spi.utils.ResourceFinder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+//TODO: Add support for Schema Registry
+//TODO: Making descriptor file accessible on all servers
+public class ProtoBufMessageDecoder implements StreamMessageDecoder<byte[]> {
+  private static final Logger LOGGER = LoggerFactory.getLogger(ProtoBufMessageDecoder.class);
+
+  public static final String DESCRIPTOR_FILE_PATH = "descriptorFile";
+  public static final String PROTO_CLASS_NAME = "protoClassName";
+
+  private DynamicMessage _dynamicMessage;
+  private ProtoBufRecordExtractor _recordExtractor;
+  private String _protoClassName;
+
+  @Override
+  public void init(Map<String, String> props, Set<String> fieldsToRead, String topicName)
+      throws Exception {
+    Preconditions.checkState(props.containsKey(DESCRIPTOR_FILE_PATH),
```

Comment:
I am not familiar with what a descriptor file represents after reading https://github.com/os72/protobuf-dynamic and viewing `sample.desc` . Is that a binary file that is generated by proto compiler?

Reply:
Yes. So basically `.proto` file contains the schema. From this schema, you can either generate java classes or you can generate `.desc` binary file.  The latter approach allows us to parse any proto message easily without relying on java impl. It is a bit slower though.

Reply:
the slowness should only be a one-time overhead right? we do not decode `desc` files on every message

Reply:
The slowness is not due to the decoding of .desc file.
It is due to the fact we are using `DynamicMessage` class instead of compiled Proto java class to deserialize the messages.

Reply:
I did some benchmark on this few years back - https://codeburst.io/using-dynamic-messages-in-protocol-buffers-in-scala-9fda4f0efcb3

Reply:
These are on my machine with proto version we are using. Significant improvments from ones in blog post.
```
Benchmark                                             (_numRecords)  Mode  Cnt   Score   Error  Units
BenchmarkDynamicMessage.compiledClassDeserialization         100000  avgt    5  12.787 ± 0.297  ms/op
BenchmarkDynamicMessage.dynamicClassDeserialization          100000  avgt    5  28.150 ± 0.691  ms/op
```